fl4health/utils/metrics.py

```diff
@@ -325,22 +357,47 @@ def __init__(self, metrics: Sequence[Metric], metric_manager_name: str) -> None:
         self.metric_manager_name = metric_manager_name
         self.metrics_per_prediction_type: Dict[str, Sequence[Metric]] = {}
 
-    def update(self, preds: Dict[str, torch.Tensor], target: torch.Tensor) -> None:
+    def update(self, preds: TorchPredType, target: TorchTargetType) -> None:
         """
         Updates (or creates then updates) a list of metrics for each prediction type.
 
         Args:
-            preds (Dict[str, torch.Tensor]): A dictionary of preds from the model
-            target (torch.Tensor): The ground truth labels for the data
+            preds (TorchPredType): A dictionary of preds from the model
+            target (TorchTargetType): The ground truth labels for the data. If
+                target is a dictionary with more than one item, then each value
+                in the preds dictionary is evaluated with the value that has
+                the same key in the target dictionary. If target has only one
+                item or is a torch.Tensor, then the same target is used for all
+                predictions
         """
         if not self.metrics_per_prediction_type:
             self.metrics_per_prediction_type = {key: copy.deepcopy(self.original_metrics) for key in preds.keys()}
 
+        # Check if there are multiple targets
```

Comment:
Not sure if there is a way around that

Reply:
This was actually a little trickier than I realized because of how we defined the types for target. There is a subtlety with the way mypy treats constrained typevars (as we are using for our typing in this PR) vs. Union typing. There is a really [nice explanation here](https://stackoverflow.com/questions/58903906/whats-the-difference-between-a-constrained-typevar-and-a-union#:~:text=So%2C%20use%20a%20TypeVar%20if,need%20to%20match%20each%20other.) The TLDR is that because we are using constrained typevars, the second that we narrow the scope of the target, it becomes fixed as that value, which means we have less flexibility in certain aspects. My suggestion is for us to actually change our type aliases to the following 
```
TorchInputType = Union[torch.Tensor, Dict[str, torch.Tensor]]
TorchTargetType = Union[torch.Tensor, Dict[str, torch.Tensor]]
```
This makes the code above work without issue and I don't think it compromises anything we were trying to do elsewhere. However, I'm happy to have a deeper discussion about this if you have a strong feeling that we shouldn't use unions.

Reply:
I actually considered this, but went with TypeVars to stay consistent. I have no objections to this

Reply:
Just to elaborate a little bit, as I understand it, the only reason to use TypeVars over unions is that it enforces that arguments of the same TypeVar within the same scope must have the same type. However I do not see any scenarios where we would have multiple TorchTargetTypes or TorchInputTypes  etc. as arguments to a method. So I think this is an unnecessary restriction and it is safe to move to unions